repo2docker/buildpacks/pyproject/__init__.py

```diff
@@ -0,0 +1,149 @@
+                        )
+
+        self._python_version = runtime_python_version
+        return self._python_version
+
+    @lru_cache
+    def get_preassemble_script_files(self):
+        """Return files needed for preassembly"""
+        files = super().get_preassemble_script_files()
+        path = self.binder_path("pyproject.toml")
+        if os.path.exists(path):
+            files[path] = path
+        return files
+
+    @lru_cache
+    def get_preassemble_scripts(self):
+        """scripts to run prior to staging the repo contents"""
+        scripts = super().get_preassemble_scripts()
+        return scripts
+
+    @lru_cache
+    def get_assemble_scripts(self):
+        """Return series of build-steps specific to this repository."""
+        # If we have pyproject.toml declare the
+        # use of Python 2, Python 2.7 will be made available in the *kernel*
+        # environment. The notebook servers environment on the other hand
+        # requires Python 3 but may require something additional installed in it
+        # still such as `nbgitpuller`. For this purpose, a "requirements3.txt"
+        # file will be used to install dependencies for the notebook servers
+        # environment, if Python 2 had been specified for the kernel
+        # environment.
+        assemble_scripts = super().get_assemble_scripts()
+
+        if self.separate_kernel_env:
+            # using legacy Python (e.g. 2.7) as a kernel
+
+            # requirements3.txt allows for packages to be installed to the
+            # notebook servers Python environment
+            nb_requirements_file = self.binder_path("requirements3.txt")
+            if os.path.exists(nb_requirements_file):
+                assemble_scripts.append(
+                    (
+                        "${NB_USER}",
+                        f'${{NB_PYTHON_PREFIX}}/bin/pip install --no-cache-dir -r "{nb_requirements_file}"',
+                    )
+                )
+
+        assemble_scripts.append(
+            (
+                "${NB_USER}",
+                "${KERNEL_PYTHON_PREFIX}/bin/pip install --no-cache-dir --editable .",
+            )
+        )
+
+        return assemble_scripts
+
+    def detect(self):
+        """Check if current repo should be built with the pyproject.toml buildpack."""
+        # first make sure python is not explicitly unwanted
+        name, _, _ = self.runtime
+        if name != "python":
+            return False
+
+        pyproject_file = self.binder_path("pyproject.toml")
+        if os.path.exists(pyproject_file):
+            with open(pyproject_file, "rb") as _pyproject_file:
+                pyproject_toml = tomllib.load(_pyproject_file)
+
+            if (
+                ("project" in pyproject_toml)
+                and ("build-system" in pyproject_toml)
+                and ("requires" in pyproject_toml["build-system"])
+            ):
+                return os.path.exists(pyproject_file)
```

Comment:
this condition is always true at this point, we can avoid checking it again:

```suggestion
                return True
```